### ISSUE TITLE
SGD

### ISSUE THREAD

**Issue by bkj** [tkipf/keras-gcn#3]
Have you ever pursued methods for training these models with batches smaller than the entire size of the data?  

I understand that making a prediction for a node involves access to it's n-th degree neighbors, but have you ever tried to implement these methods in practice and/or are you aware of pitfalls involved?

**Comment by tkipf** [comment]
Excellent point. While naive mini-batch stochastic gradient descent as you described it is definitely possible for these kinds of models, I have not yet implemented such a variant as it comes with a number of potential issues: taking a mini-batch of nodes and their complete k-th-order neighborhoods (for a model with k layers) potentially already includes the full graph and therefore every single node in the graph might have to be kept in memory for a single mini-batch update. This is especially true for scale-free networks (most real-world graph datasets show such a power-law behavior). Thereby, naive SGD cannot be used to tackle the memory issue of the model, unfortunately.

**Comment by bkj** [comment]
Interesting.  I have some use cases where these methods seem very useful, but they're probably a little large for the current implementation. (1.6M nodes and 30M edges for the smallest one).  I'll do some experiments and report back if I (don't) find anything interesting.

**Comment by tkipf** [comment]
See here for a mini-batch solution for this class of models: https://github.com/williamleif/GraphSAGE
